include/graph_contracts.hpp

```diff
@@ -0,0 +1,444 @@
+};
+static_assert(Port<dyn_port>);
+
+class block { // TODO: test block for development purposes to be merged/replaced with impl in graph.hpp
+    using setting_map = std::map<std::string, int, std::less<>>;
+
+    std::string _name;
+    std::vector<dyn_port> _input_ports{};
+    std::vector<dyn_port> _output_ports{};
+    setting_map _exec_metrics{}; //  →  std::map<string, pmt> → fair scheduling, 'int' stand-in for pmtv
+
+public:
+    explicit block(std::string_view name) : _name(name) {};
+
+    [[nodiscard]] std::string_view name() const noexcept { return _name; }
+    void work() noexcept { /* to be implemented in derived classes */ }
+    [[nodiscard]] std::vector<dyn_port>const & input_ports() const noexcept { return _input_ports; }
+    [[nodiscard]] std::vector<dyn_port>& input_ports() noexcept { return _input_ports; }
+    [[nodiscard]] std::vector<dyn_port>const & output_ports() const noexcept { return _output_ports; }
+    [[nodiscard]] std::vector<dyn_port>& output_ports() noexcept { return _output_ports; }
+
+    [[nodiscard]] setting_map &exec_metrics() noexcept { return _exec_metrics; }
+    [[nodiscard]] setting_map const &exec_metrics() const noexcept { return _exec_metrics; }
+
+//    { t.connect(src_port, dst_block, dst_port) } -> std::same_as<std::vector<connection_result_t>>;
+};
+static_assert(Block<block>);
+
+
+class edge {
+    block& _src_block;
+    dyn_port& _src_port;
+    block& _dst_block;
+    dyn_port& _dst_port;
+    int32_t _weight;
+    std::string _name; // custom edge name
+    bool _connected;
+
+public:
+    edge() = delete;
+    edge& operator=(const edge&) = delete;
+    edge(block& src_block, dyn_port& from, block& dst_block, dyn_port& to, int32_t weight, std::string_view name) :
+        _src_block(src_block), _src_port(from), _dst_block(dst_block), _dst_port(to), _weight(weight), _name(name) {
+        if (from.pmt_type().index() != to.pmt_type().index()) {
+            throw fmt::format("edge({}::{}<{}> -> {}::{}<{}>, weight: {}, name:\"{}\") incompatible to type id='{}'",
+                _src_block.name(), _src_port.name(), _src_port.pmt_type().index(),
+                _dst_block.name(), _dst_port.name(), _dst_port.pmt_type().index(),
+                _weight, _name, _dst_port.pmt_type().index());
+        }
+    }
+
+    [[nodiscard]] constexpr int32_t weight() const noexcept { return _weight; }
+    [[nodiscard]] constexpr std::string_view name() const noexcept { return _name; }
+    [[nodiscard]] constexpr bool connected() const noexcept { return _connected; }
+    [[nodiscard]] connection_result_t connect() const noexcept { return connection_result_t::FAILED; }
+    [[nodiscard]] connection_result_t disconnect() const noexcept { return _dst_port.disconnect(); }
+};
+
+[[nodiscard]] connection_result_t add_edge(std::vector<edge>& edges, block& src_block, dyn_port& src_port, block& dst_block, dyn_port& dst_port, int32_t weight = 0, std::string_view name = "unnamed edge") noexcept {
```

Comment:
Similar comment to above - edges should not be in a plain vector, IMO this needs a class with a proper API

Reply:
done.